yt/yql/agent/yql_agent.cpp

```diff
@@ -32,6 +33,70 @@ const auto& Logger = YqlAgentLogger;
 
 ////////////////////////////////////////////////////////////////////////////////
 
+static std::optional<TString> TryIssueToken(const TQueryId queryId, const TString& user, const std::vector<TString>& clusters, THashMap<TString, NApi::NNative::IClientPtr>& queryClients, const TDuration& expirationTimeout)
+{
+    TString token;
+    if (clusters.empty()) {
+        return token;
+    }
+
+    auto options = NApi::TIssueTemporaryTokenOptions{ .ExpirationTimeout = expirationTimeout };
+    auto attributes = CreateEphemeralAttributes();
+    attributes->Set("query_id", queryId);
+    attributes->Set("responsible", "query_tracker");
+
+    for (auto& cluster : clusters) {
+        YT_LOG_DEBUG("Requesting token (User: %v, Cluster: %v)", user, cluster);
+        auto rspOrError = token.empty()
+            ? WaitFor(queryClients[cluster]->IssueTemporaryToken(user, attributes, options))
+            : WaitFor(queryClients[cluster]->IssueSpecificTemporaryToken(user, token, attributes, options));
+
+        if (!rspOrError.IsOK()) {
+            YT_LOG_WARNING("Token request failed (User: %v, Cluster: %v)", user, cluster);
+            if (rspOrError.FindMatching(NYTree::EErrorCode::AlreadyExists)) {
+                YT_LOG_WARNING("Requested token already exists in the cluster (User: %v, Cluster: %v)", user, cluster);
+                return std::nullopt;
+            }
+            rspOrError.ThrowOnError();
+        }
+
+        if (token.empty()) {
+            token = rspOrError.ValueOrThrow().Token;
+        }
+        YT_LOG_DEBUG("Token received (User: %v, Cluster: %v)", user, cluster);
+    }
+
+    return token;
+}
+
+static TString IssueToken(const TQueryId queryId, const TString& user, const std::vector<TString>& clusters, THashMap<TString, NApi::NNative::IClientPtr>& queryClients, const TDuration& expirationTimeout)
+{
+    while (true) {
```

Comment:
Let's do it 10 times, for example, and throw an error on failure to issue token. It will be much easier to debug in case some issue happens with token generation.